functions/index.js

```diff
@@ -0,0 +1,208 @@
+        const songDocRef = db.collection("songs").doc(songId);
+
+        console.log("GET WARPED SONG");
+
+        let responseObj = {};
+
+        songDocRef
+            .get()
+            .then(async doc => {
+                console.log("RESOLVED REQUEST FOR DOC");
+
+                // if song with matching id exists
+                if (doc.exists) {
+                    console.log("DOC EXISTS");
+
+                    // if warped version exists.
+                    const { original, warped, name } = doc.data();
+                    if (warped) {
+                        console.log("HAS WARPED");
+                        // return it
+                        res.send({ name, warped });
+                        return;
+                    } else if (original) {
+                        // translate to warped version
+                        const warped = await translateSongLyrics(original);
+                        // store in firebase
+                        songDocRef.set({
+                            name,
+                            original,
+                            warped
+                        });
+                        // return it
+                        res.send({
+                            name,
+                            warped
+                        });
+                        return;
+                    }
+                } else {
+                    console.log("DOC DOES NOT EXIST");
+
+                    // else
+                    // scrape it to get original
+                    const { data } = await proxyRequest(`/songs/${songId}`, {});
+
+                    const song = data.response.song;
+                    console.log("GOT DATA");
+
+                    const path = song.path;
+                    const originalLyrics = await scrapeContent(path);
+                    console.log("ORIGINAL LYRICS");
+
+                    // translate to warped version
+                    const warpedLyrics = await translateSongLyrics(
+                        originalLyrics
+                    );
+                    console.log("WARPED LYRICS", warpedLyrics);
+
+                    // store in firebase
+                    try {
+                        songDocRef.set({
+                            name: song.primary_artist.name,
+                            original: originalLyrics,
+                            warped: warpedLyrics
+                        });
+                    } catch (err) {
+                        console.log(err);
+                    }
+                    // return it
+                    res.send({
+                        name: song.primary_artist.name,
+                        warped: warpedLyrics
+                    });
+                    return;
+                }
+                return;
+            })
+            .catch(err => {
+                res.send(err);
+            });
+    });
+});
+
+const proxy = functions.https.onRequest(async (req, res) => {
```

Comment:
This is the new proxy endpoint function. I actually pulled the proxy logic itself into a shareable function so that other Firebase Functions can make use of it.